pkg/util/serialization/deserialization_util.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serialization
+
+import (
+	"bytes"
+	"time"
+	"unsafe"
+
+	"github.com/pingcap/tidb/pkg/types"
+)
+
+func deserializeLength(buf []byte, pos *int64) int {
```

Comment:
comment from chatgpt:
You can simplify the function by only returning the buffer:

```go
// deserializeBuffer deserializes a buffer.
func deserializeBuffer(buf []byte, pos *int64) []byte {
	length := DeserializeInt(buf, pos)
	data := make([]byte, length)
	copy(data, buf[*pos:*pos+length])
	*pos += length
	return data
}
```

And then use it in your other deserialization functions:

```go
// DeserializeOpaque deserializes Opaque type
func DeserializeOpaque(buf []byte, pos *int64) types.Opaque {
	retVal := types.Opaque{}
	retVal.TypeCode = buf[*pos]
	*pos++
	retVal.Buf = deserializeBuffer(buf, pos)
	return retVal
}

// DeserializeString deserializes String type
func DeserializeString(buf []byte, pos *int64) string {
	return string(deserializeBuffer(buf, pos))
}

// DeserializeBytesBuffer deserializes bytes.Buffer type
func DeserializeBytesBuffer(buf []byte, pos *int64) *bytes.Buffer {
	tmp := deserializeBuffer(buf, pos)
	return bytes.NewBuffer(tmp)
}

// ... (other deserialization functions)
```

Reply:
> comment from chatgpt: You can simplify the function by only returning the buffer:
> 
> ```go
> // deserializeBuffer deserializes a buffer.
> func deserializeBuffer(buf []byte, pos *int64) []byte {
> 	length := DeserializeInt(buf, pos)
> 	data := make([]byte, length)
> 	copy(data, buf[*pos:*pos+length])
> 	*pos += length
> 	return data
> }
> ```
> 
> And then use it in your other deserialization functions:
> 
> ```go
> // DeserializeOpaque deserializes Opaque type
> func DeserializeOpaque(buf []byte, pos *int64) types.Opaque {
> 	retVal := types.Opaque{}
> 	retVal.TypeCode = buf[*pos]
> 	*pos++
> 	retVal.Buf = deserializeBuffer(buf, pos)
> 	return retVal
> }
> 
> // DeserializeString deserializes String type
> func DeserializeString(buf []byte, pos *int64) string {
> 	return string(deserializeBuffer(buf, pos))
> }
> 
> // DeserializeBytesBuffer deserializes bytes.Buffer type
> func DeserializeBytesBuffer(buf []byte, pos *int64) *bytes.Buffer {
> 	tmp := deserializeBuffer(buf, pos)
> 	return bytes.NewBuffer(tmp)
> }
> 
> // ... (other deserialization functions)
> ```

I think constructing string with `deserializeBuffer` will cause useless copy.

Reply:
> comment from chatgpt: You can simplify the function by only returning the buffer:
> 
> ```go
> // deserializeBuffer deserializes a buffer.
> func deserializeBuffer(buf []byte, pos *int64) []byte {
> 	length := DeserializeInt(buf, pos)
> 	data := make([]byte, length)
> 	copy(data, buf[*pos:*pos+length])
> 	*pos += length
> 	return data
> }
> ```
> 
> And then use it in your other deserialization functions:
> 
> ```go
> // DeserializeOpaque deserializes Opaque type
> func DeserializeOpaque(buf []byte, pos *int64) types.Opaque {
> 	retVal := types.Opaque{}
> 	retVal.TypeCode = buf[*pos]
> 	*pos++
> 	retVal.Buf = deserializeBuffer(buf, pos)
> 	return retVal
> }
> 
> // DeserializeString deserializes String type
> func DeserializeString(buf []byte, pos *int64) string {
> 	return string(deserializeBuffer(buf, pos))
> }
> 
> // DeserializeBytesBuffer deserializes bytes.Buffer type
> func DeserializeBytesBuffer(buf []byte, pos *int64) *bytes.Buffer {
> 	tmp := deserializeBuffer(buf, pos)
> 	return bytes.NewBuffer(tmp)
> }
> 
> // ... (other deserialization functions)
> ```

Only returning buf is shallow copy, however sometime we need deep copy.